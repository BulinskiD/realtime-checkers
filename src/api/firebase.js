import app from "firebase/app";
import "firebase/storage";
import "firebase/auth";
import "firebase/firestore";

var firebaseConfig = {
  apiKey: "AIzaSyD3uWma_iug5o_0w6EmchIXp4qd054HqcE",
  authDomain: "realtime-checkers.firebaseapp.com",
  databaseURL: "https://realtime-checkers.firebaseio.com",
  projectId: "realtime-checkers",
  storageBucket: "realtime-checkers.appspot.com",
  messagingSenderId: "100295366319",
  appId: "1:100295366319:web:ec1c182d83b660e8"
};

// Initialize Firebase
app.initializeApp(firebaseConfig);

export const auth = app.auth();
export const firestore = app.firestore();
