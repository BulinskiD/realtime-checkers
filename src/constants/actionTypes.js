export const TEST_ACTION = 'TEST_ACTION';
export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
