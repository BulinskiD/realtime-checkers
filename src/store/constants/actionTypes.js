export const START_LOADING = 'START_LOADING';
export const STOP_LOADING = 'STOP_LOADING';
export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
export const LOGIN_FAILED = 'LOGIN_FAILED';
export const LOGOUT_SUCCESS = 'LOGOUT_SUCCESS';
export const LOGOUT_FAILED = 'LOGOUT_FAILED';
