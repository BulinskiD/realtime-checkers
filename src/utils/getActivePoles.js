export default (selectedChecker, checkersPosition) => {
    const {col, row, color} = selectedChecker;
    const availablePoles = [];
    let availableRow;
    let nextAvailableRow;

    if(color === 'black') {
        availableRow = row - 1;
        nextAvailableRow = row -2;
    } else {
        availableRow = row + 1;
        nextAvailableRow = row +2;
    }

    if(checkersPosition.filter(item => item.col === col + 1 && item.row === availableRow).length === 0)
        availablePoles.push({col: col+1, row: availableRow});
    else if(checkersPosition.filter(item => item.col === col + 2 && item.row === nextAvailableRow).length === 0)
        availablePoles.push({col: col+2, row: nextAvailableRow});

    if(checkersPosition.filter(item => item.col === col - 1 && item.row === availableRow).length === 0)
        availablePoles.push({col: col-1, row: availableRow});
    else if(checkersPosition.filter(item => item.col === col - 2 && item.row === nextAvailableRow).length === 0)
        availablePoles.push({col: col-2, row: nextAvailableRow});

    return availablePoles;
}