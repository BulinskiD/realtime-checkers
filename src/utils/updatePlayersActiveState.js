import { firestore } from "../api/firebase";

export default async (players, email, gameID, isMounting) => {
  let isChange = false;
  const newPlayers = players.map(item => {
    if (item.email === email && item.active !== isMounting) {
      item.active = isMounting;
      isChange = true;
    }
    return item;
  });

  if (isChange)
    await firestore
      .collection("games")
      .doc(gameID)
      .update({
        players: newPlayers,
        updated: Date.now()
      });
};
