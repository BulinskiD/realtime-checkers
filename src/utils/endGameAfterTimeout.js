import { WHITE_WINNER, BLACK_WINNER } from "../store/constants/actionTypes";
import { firestore } from "../api/firebase";

export default async (time, status, players, gameID, inter) => {
  if (time > 180 && (status === "white" || status === "black")) {
    clearInterval(inter);
    const winnerStatus = status === "white" ? "black" : "white";

    if (players.filter(item => item.active).length !== 0)
      await firestore
        .collection("games")
        .doc(gameID)
        .update({
          status: winnerStatus === "white" ? WHITE_WINNER : BLACK_WINNER,
          updated: Date.now()
        });
  }
};
