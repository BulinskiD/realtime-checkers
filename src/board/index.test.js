import React from "react";
import { act } from "react-dom/test-utils";
import ConnectedBoard, { Board } from "./index";
import configureMockStore from "redux-mock-store";
import { shallow, mount } from "enzyme";
import { Provider } from "react-redux";
import { firestore } from "../api/firebase";
import getActivePoles from "../utils/getActivePoles";
import PlayersManager from "./playersManager";
import updatePlayersActiveState from "../utils/updatePlayersActiveState";
import endGameAfterTimeout from "../utils/endGameAfterTimeout";
import { BLACK_WINNER } from "../store/constants/actionTypes";

jest.mock("../utils/endGameAfterTimeout");
jest.mock("../utils/updatePlayersActiveState");
jest.mock("./playersManager");
jest.mock("../utils/getActivePoles");
jest.mock("../api/firebase");
const item = { id: "22", data: () => data };
const unsubscriber = jest.fn();
const onSnapshot = jest.fn(func => {
  func(item);
  return unsubscriber;
});
const doc = jest.fn();
firestore.collection = jest.fn();
firestore.collection.mockReturnValue({ doc });
doc.mockReturnValue({ onSnapshot });

jest.mock("../utils/startGame");
const setNewGameState = jest.fn();
const setActivePoles = jest.fn();
const selectGame = jest.fn();

const mockStore = configureMockStore();

let data;
getActivePoles.mockReturnValue({ availablePoles: { col: 2, row: 2 } });

describe("Board", () => {
  beforeEach(() => {
    setNewGameState.mockClear();
    setActivePoles.mockClear();
    selectGame.mockClear();
    unsubscriber.mockClear();
    endGameAfterTimeout.mockClear();
    data = {
      currentGame: {
        status: "white",
        selectedChecker: { col: 2, row: 2 },
        checkersPosition: [{ col: 2, row: 2, color: "black", selected: false }],
        nextMove: false,
        players: [{ test: test }, { test: test }],
        updated: 100
      },
      user: { email: "test" },
      setNewGameState,
      selectGame,
      setActivePoles,
      match: { params: { id: "22" } }
    };
  });

  it("should render 64 poles and Start game button for given props", () => {
    const props = {
      ...data
    };
    const component = shallow(<Board {...props} />);
    expect(component.find("Connect(Pole)").length).toBe(64);
    expect(component).toMatchSnapshot();
  });

  it("should call firestore subscribe on component load", () => {
    const store = mockStore(data);
    const props = {
      ...data
    };

    mount(
      <Provider store={store}>
        <Board {...props} />
      </Provider>
    );

    expect(selectGame).toHaveBeenCalledWith("22");
    expect(setNewGameState).toHaveBeenCalledWith("22", data);
  });

  it("should unsubscribe on component unmount", () => {
    const store = mockStore(data);
    const clearCurrentGame = jest.fn();
    const props = {
      ...data,
      clearCurrentGame
    };
    const component = mount(
      <Provider store={store}>
        <Board {...props} />
      </Provider>
    );

    component.unmount();
    expect(unsubscriber).toHaveBeenCalledTimes(1);
    expect(clearCurrentGame).toHaveBeenCalledTimes(1);
  });

  it("should call setActivePoles when selectedChecker changes", () => {
    const store = mockStore(data);
    const props = {
      ...data
    };
    mount(
      <Provider store={store}>
        <Board {...props} />
      </Provider>
    );

    expect(setActivePoles).toBeCalledWith({ col: 2, row: 2 });
  });

  it("should call endGameAfterTimeout when updated changes", () => {
    Math.ceil = jest.fn().mockReturnValue(300);
    const store = mockStore({ ...data });
    jest.useFakeTimers();
    const props = {
      ...data
    };

    act(() => {
      mount(
        <Provider store={store}>
          <Board {...props} />
        </Provider>
      );
      jest.runOnlyPendingTimers();
    });

    expect(endGameAfterTimeout).toBeCalledWith(
      300,
      "white",
      data.currentGame.players,
      "22",
      2 //TODO Check why timer ref equals 2
    );
  });

  it("should call setTimeSinceMove with ended when game is ended", () => {
    data.currentGame.status = BLACK_WINNER;
    const store = mockStore({ ...data });
    const props = {
      ...data
    };

    let component = {};
    act(() => {
      component = mount(
        <Provider store={store}>
          <Board {...props} />
        </Provider>
      );
    });
    expect(component.find(".time").text()).toBe("ended");
  });

  it("shouldn't call setActivePoles when selectedChecker is null", () => {
    const store = mockStore(data);
    data.currentGame.selectedChecker = null;
    const props = {
      ...data
    };
    mount(
      <Provider store={store}>
        <Board {...props} />
      </Provider>
    );

    expect(setActivePoles).toBeCalledTimes(0);
  });
});

describe("MapStateToProps", () => {
  it("should correctly map state to props", () => {
    const store = mockStore({
      currentGame: {
        status: "white",
        selectedChecker: {},
        checkersPosition: [{ col: 2, row: 2, color: "black", selected: false }],
        players: [{ email: "test", color: "white" }],
        nextMove: false
      },
      user: { email: "test" }
    });

    const board = shallow(
      <ConnectedBoard store={store} {...{ match: { param: { id: 2 } } }} />
    );

    expect(board.find("Board").props().currentGame).toStrictEqual({
      status: "white",
      players: [{ email: "test", color: "white" }],
      selectedChecker: {},
      checkersPosition: [{ col: 2, row: 2, color: "black", selected: false }],
      nextMove: false
    });
    expect(board.find("Board").props().isActiveTurn).toBe(true);
  });
});
