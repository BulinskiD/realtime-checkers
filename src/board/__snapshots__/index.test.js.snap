// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board should render 64 poles and Start game button for given props 1`] = `
<styled.div>
  <div
    className="time"
  />
  <Connect(PlayersManager)
    gameID="22"
  />
  <styled.div>
    <Connect(Pole)
      col={0}
      key="0/0"
      row={0}
    />
    <Connect(Pole)
      col={1}
      key="0/1"
      row={0}
    />
    <Connect(Pole)
      col={2}
      key="0/2"
      row={0}
    />
    <Connect(Pole)
      col={3}
      key="0/3"
      row={0}
    />
    <Connect(Pole)
      col={4}
      key="0/4"
      row={0}
    />
    <Connect(Pole)
      col={5}
      key="0/5"
      row={0}
    />
    <Connect(Pole)
      col={6}
      key="0/6"
      row={0}
    />
    <Connect(Pole)
      col={7}
      key="0/7"
      row={0}
    />
    <Connect(Pole)
      col={0}
      key="1/0"
      row={1}
    />
    <Connect(Pole)
      col={1}
      key="1/1"
      row={1}
    />
    <Connect(Pole)
      col={2}
      key="1/2"
      row={1}
    />
    <Connect(Pole)
      col={3}
      key="1/3"
      row={1}
    />
    <Connect(Pole)
      col={4}
      key="1/4"
      row={1}
    />
    <Connect(Pole)
      col={5}
      key="1/5"
      row={1}
    />
    <Connect(Pole)
      col={6}
      key="1/6"
      row={1}
    />
    <Connect(Pole)
      col={7}
      key="1/7"
      row={1}
    />
    <Connect(Pole)
      col={0}
      key="2/0"
      row={2}
    />
    <Connect(Pole)
      col={1}
      key="2/1"
      row={2}
    />
    <Connect(Pole)
      col={2}
      key="2/2"
      row={2}
    />
    <Connect(Pole)
      col={3}
      key="2/3"
      row={2}
    />
    <Connect(Pole)
      col={4}
      key="2/4"
      row={2}
    />
    <Connect(Pole)
      col={5}
      key="2/5"
      row={2}
    />
    <Connect(Pole)
      col={6}
      key="2/6"
      row={2}
    />
    <Connect(Pole)
      col={7}
      key="2/7"
      row={2}
    />
    <Connect(Pole)
      col={0}
      key="3/0"
      row={3}
    />
    <Connect(Pole)
      col={1}
      key="3/1"
      row={3}
    />
    <Connect(Pole)
      col={2}
      key="3/2"
      row={3}
    />
    <Connect(Pole)
      col={3}
      key="3/3"
      row={3}
    />
    <Connect(Pole)
      col={4}
      key="3/4"
      row={3}
    />
    <Connect(Pole)
      col={5}
      key="3/5"
      row={3}
    />
    <Connect(Pole)
      col={6}
      key="3/6"
      row={3}
    />
    <Connect(Pole)
      col={7}
      key="3/7"
      row={3}
    />
    <Connect(Pole)
      col={0}
      key="4/0"
      row={4}
    />
    <Connect(Pole)
      col={1}
      key="4/1"
      row={4}
    />
    <Connect(Pole)
      col={2}
      key="4/2"
      row={4}
    />
    <Connect(Pole)
      col={3}
      key="4/3"
      row={4}
    />
    <Connect(Pole)
      col={4}
      key="4/4"
      row={4}
    />
    <Connect(Pole)
      col={5}
      key="4/5"
      row={4}
    />
    <Connect(Pole)
      col={6}
      key="4/6"
      row={4}
    />
    <Connect(Pole)
      col={7}
      key="4/7"
      row={4}
    />
    <Connect(Pole)
      col={0}
      key="5/0"
      row={5}
    />
    <Connect(Pole)
      col={1}
      key="5/1"
      row={5}
    />
    <Connect(Pole)
      col={2}
      key="5/2"
      row={5}
    />
    <Connect(Pole)
      col={3}
      key="5/3"
      row={5}
    />
    <Connect(Pole)
      col={4}
      key="5/4"
      row={5}
    />
    <Connect(Pole)
      col={5}
      key="5/5"
      row={5}
    />
    <Connect(Pole)
      col={6}
      key="5/6"
      row={5}
    />
    <Connect(Pole)
      col={7}
      key="5/7"
      row={5}
    />
    <Connect(Pole)
      col={0}
      key="6/0"
      row={6}
    />
    <Connect(Pole)
      col={1}
      key="6/1"
      row={6}
    />
    <Connect(Pole)
      col={2}
      key="6/2"
      row={6}
    />
    <Connect(Pole)
      col={3}
      key="6/3"
      row={6}
    />
    <Connect(Pole)
      col={4}
      key="6/4"
      row={6}
    />
    <Connect(Pole)
      col={5}
      key="6/5"
      row={6}
    />
    <Connect(Pole)
      col={6}
      key="6/6"
      row={6}
    />
    <Connect(Pole)
      col={7}
      key="6/7"
      row={6}
    />
    <Connect(Pole)
      col={0}
      key="7/0"
      row={7}
    />
    <Connect(Pole)
      col={1}
      key="7/1"
      row={7}
    />
    <Connect(Pole)
      col={2}
      key="7/2"
      row={7}
    />
    <Connect(Pole)
      col={3}
      key="7/3"
      row={7}
    />
    <Connect(Pole)
      col={4}
      key="7/4"
      row={7}
    />
    <Connect(Pole)
      col={5}
      key="7/5"
      row={7}
    />
    <Connect(Pole)
      col={6}
      key="7/6"
      row={7}
    />
    <Connect(Pole)
      col={7}
      key="7/7"
      row={7}
    />
  </styled.div>
</styled.div>
`;
